db/model/Gdoc/GdocFactory.ts

```diff
@@ -1,154 +1,556 @@
+    )
+    return gdocs as GdocAuthor[]
+}
+
+export async function getAndLoadListedGdocPosts(
+    knex: KnexReadonlyTransaction
+): Promise<GdocPost[]> {
+    // TODO: Check if we shouldn't also restrict the types of gdocs here
+    const rows = await knexRaw<DbRawPostGdoc>(
+        knex,
+        `-- sql
+            SELECT *
+            FROM posts_gdocs
+            WHERE published = 1
+            publicationContext: ${OwidGdocPublicationContext.listed}
+            publishedAt <= NOW()
+            ORDER BY publishedAt DESC`,
+        {}
+    )
+    const ids = rows.map((row) => row.id)
+    const tags = await knexRaw<DbPlainTag>(
+        knex,
+        `-- sql
+                SELECT gt.gdocId as gdocId, tags.*
+                FROM tags
+                JOIN posts_gdocs_x_tags gt ON gt.tagId = tags.id
+                WHERE gt.gdocId in (:ids)`,
+        { ids: ids }
+    )
+    const groupedTags = groupBy(tags, "gdocId")
+    const enrichedRows = rows.map((row) => {
+        return {
+            ...parsePostsGdocsRow(row),
+            tags: groupedTags[row.id] ? groupedTags[row.id] : null,
+        } satisfies OwidGdocBaseInterface
+    })
+    const gdocs = await Promise.all(
+        enrichedRows.map(async (row) => loadGdocFromGdocBase(knex, row))
+    )
+    return gdocs as GdocPost[]
+}
 
-        await gdoc.loadState()
+export async function setTagsForGdoc(
+    knex: KnexReadWriteTransaction,
+    gdocId: string,
+    tagIds: Pick<DbPlainTag, "id">[]
+): Promise<void> {
+    await knex.table(PostsGdocsXTagsTableName).where("gdocId", gdocId).delete()
+    await knex
+        .table(PostsGdocsXTagsTableName)
+        .insert(tagIds.map(({ id: tagId }) => ({ gdocId, tagId })))
+}
 
-        return gdoc
+export function getDbEnrichedGdocFromOwidGdoc(
+    gdoc: OwidGdoc | GdocBase
+): DbEnrichedPostGdoc {
+    const enrichedGdoc = {
+        breadcrumbs: gdoc.breadcrumbs,
+        content: gdoc.content,
+        createdAt: gdoc.createdAt,
+        id: gdoc.id,
+        markdown: gdoc.markdown,
+        publicationContext: gdoc.publicationContext,
+        published: gdoc.published,
+        publishedAt: gdoc.publishedAt,
+        revisionId: gdoc.revisionId,
+        slug: gdoc.slug,
+        updatedAt: gdoc.updatedAt,
+    } satisfies DbEnrichedPostGdoc
+    return enrichedGdoc
+}
+export async function upsertGdoc(
+    knex: KnexReadWriteTransaction,
+    gdoc: OwidGdoc | GdocBase
+): Promise<number[]> {
+    let sql = undefined
+    try {
+        const enrichedGdoc = getDbEnrichedGdocFromOwidGdoc(gdoc)
+        const rawPost = serializePostsGdocsRow(enrichedGdoc)
+        const query = knex
+            .table(PostsGdocsTableName)
+            .insert(rawPost)
+            .onConflict("id")
+            .merge()
+        sql = query.toSQL()
+        return query
+    } catch (e) {
+        console.error(`Error occured in sql: ${sql}`, e)
+        throw e
     }
 }
+
+// TODO:
```

Comment:
TODO?

Reply:
I like sprinkling mystery TODOs in a codebase as I hack through it ...